Animal User Story

Mononofuu opened the pull request:
- Animal class, Dto, table  reworked
- Species class, Dto, table  reworked
- JdbcTemplate persistence provider  added
- Animal, Species  and House specifications added
- House capacity map initialization  added to House Repository
- House and Animal business logic in services
- House exception
- Animal exception

Mononofuu committed:
Done:
 - Animal class, Dto, table  reworked
 - Species class, Dto, table  reworked
 - JdbcTemplate persistence provider  added
 - Animal, Species  and House specifications added
 - House capacity map initialization  added to House Repository
 - House and Animal business logic in services
 - House exception
 - Animal exception

skyfenko commented on persistence/src/main/java/edu/softserve/zoo/persistence/provider/JdbcTemplatePersistenceProvider.java:
```diff
@@ -0,0 +1,25 @@
+     * @return map of Key-Values pairs
+     * @throws PersistenceProviderException if the received specification is null
+     */
+    <K extends Number, V extends Number> Map<K, V> getMap(SqlGetMapSpecification<K, V> specification);
```
why do you narrow generics to Numbers? what's for?

Mononofuu replied:
Maybe the main reason - is that I consider this functional as a tool to return the ratio of the result of the aggregate function to id. Can you give me some advice on this?

skyfenko commented on persistence/src/main/java/edu/softserve/zoo/persistence/repository/impl/HouseRepositoryImpl.java:
```diff
@@ -1,16 +1,39 @@
+
+    @PostConstruct
+    public void initHouseCapacityMap() {
+        houseCapacityMap = Collections.synchronizedMap(persistenceProvider.getMap(new HouseGetCapacityMapSpecification()));
```
bad practice to use collections.syncMap. look at thread safe variants

skyfenko replied:
why do you need sync map?

skyfenko commented on persistence/src/main/java/edu/softserve/zoo/persistence/specification/hibernate/impl/house/HouseGetAllBySpeciesIdSpecification.java:
```diff
@@ -0,0 +1,35 @@
+     * @param speciesId {@link Species} identifier to filter houses
+     */
+    public HouseGetAllBySpeciesIdSpecification(Long speciesId) {
+        this.speciesId = speciesId;
```
ok. assume this id is null. what's next?

Mononofuu replied:
validation was moved to the constructor, so exception will be thrown at this stage

Mononofuu committed:
some refactor:
 - specification validation moved to constructor
 - concurrent hashMap used instead of synchronizedMap
